src/utils/graph/constraints.js

```diff
@@ -0,0 +1,106 @@
+import { distance1d, greaterOrEqual, equalTo, subtract } from './common';
+
+/**
+ * Constraint base definitions.
+ *
+ * See `solve` function of `solver.js` for constraint specification.
+ */
+
+/**
+ * Layout constraint in Y for separating rows
+ */
+export const rowConstraint = {
+  property: 'y',
+  difference: subtract,
```

Comment:
Yeah I see what you mean for sure, there's a few reasons for this though:

- Future constraint improvements will potentially need this option
- The solver is intentionally generalised and therefore makes no assumptions
- Supporting default values was removed in this PR in favour of performance (it's significant)
- I think this is actually clearer to read, being explicit rather than implicit is useful here

But I'll update the description with specific note on this.

Reply:
ok cheers!